contributors/design-proposals/scheduling/coscheduling.md

```diff
@@ -0,0 +1,174 @@
+
+### Customized Controller
+
+A typical example of customized controller is [kubeflow/tf-operator](https://github.com/kubeflow/tf-operator), which managed the Pods for TensorFlow on Kubernetes, required `gang-scheduling` in upstream. Here's an example of customized controller that demonstrated the usage of  `gang-scheduling` in `kube-arbitrator`.
+
+Usually, CRD ([CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)) feature is used to introduce a customized **Kind**, named `CRDJob` as example. The customized controller, named `CRDJobController`, watches it and manages lifecycle of it:
+
+1. For each `CRDJob`, `CRDJobController` creates a `CRDJob` and `PodGroup` (one `CRDJob` with one `PodGroup` as example). The attributes of `PodGroup`should be set accordingly, e.g `numMember`; it's up to customized controller on how to manage relationship between `PodGroup` and `CRDJob`, e.g. `metadata.name`.
+2. When `CRDJobController` create Pods, its annotation should be set accordingly. `kube-arbitrator` follows gang-scheduling logic to schedule those pods in batch.
+3. When pods failed/deleted/unschedulable, it up to `CRDJobController` on how to manage `CRDJob`'s lifecycle. For example, if `CRDJobController` manages lifecycle itself, set `.spec.Policy` of `PodGroup` to nil; otherwise, `PodGroupController` will manage the lifecycle as described above.
+4. If `CRDJob` was deleted, the `PodGroup` must be deleted accordingly.
+
+## Feature Interaction
+
+### Multi-scheduler
+
+When multi-scheduler was enabled, there maybe decision conflict between different schedulers; and the kubelet will reject one pod (failed) if conflict. The controller will handle rejected pods based on its lifecycle policy for failed pods. It up to cluster admin to avoid such kind of conflict, e.g. node affinity which is supported by kube-arbitrator in 1.13.
+
+### Priority/Preemption
+
+As described in `multi-scheduler` section, it's better to run batch and service workload in different zone to avoid confliction.
+Rejected a batch/run-to-complete pod may trigger the restarting of the whole `PodGroup` which impact the performance. The solution on how to handle confliction better will be proposed in coming release.
+
+The default scheduler should also consider `PodGroup` when preempting pods, similar to `PodDisruptionBudgets`.
+
+### Pod RestartPolicy
+
+Pod's `RestartPolicy` still works as before. But for batch/run-to-compelete workload, it's better to set `RestartPolicy` to `Never` to avoid endless restart loop.
+
+### Admission Controller
+
+Because of AdmissionController, e.g. `ResourceQuota`, a few pods of a gang maybe created, but the rest fail admission pods are not created. `.spec.Resource` is introduced in `PodGroup` to reserve resources for PodGroup by `ResourceQuota`, so it makes sure the Pods of PodGroup has enough quota to create it.
```

Comment:
`spec.Resource` is specifically added to address the `ResourceQuota` admission check. `spec.Resource` is not enough to address all challenges of admitting batches of pods. So, I would rephrase this as:

If quota runs out in the middle of creating a group of pods, a few members of a PodGroup may be created, while the rest will be denied by the `ResourceQuota` admission controller. `.spec.Resource` is added in `PodGroup` to address this problem. When a `PodGroup` is created with `.spec.Resource`, so much quota is reserved for the group if there is available quota. Pods of group use the already reserved quota. By setting `.spec.Resource` properly, one can ensure that Pods of a PodGroup have enough quota at creation time.

Reply:
What happens if a pod has to be vertically scaled up (this is a WIP AFAIK)? I'd assume the podGroup has to be scaled up first successfully, before attempting to scale up pods in it. Will that be supported?

Reply:
And will scale down's be supported at all?